sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -193,7 +193,21 @@
               bundleFinalizer);
 
       // Register the appropriate handlers.
-      startFunctionRegistry.register(pTransformId, runner::startBundle);
+      switch (pTransform.getSpec().getUrn()) {
+        case PTransformTranslation.PAR_DO_TRANSFORM_URN:
+        case PTransformTranslation.SPLITTABLE_PROCESS_ELEMENTS_URN:
+        case PTransformTranslation.SPLITTABLE_PROCESS_SIZED_ELEMENTS_AND_RESTRICTIONS_URN:
+          startFunctionRegistry.register(pTransformId, runner::startBundle);
+          break;
+        case PTransformTranslation.SPLITTABLE_PAIR_WITH_RESTRICTION_URN:
```

Comment:
We can't re-use the existing `startBundle`/`finishBundle` methods since it would be confusing to the user as to which context they are executing in (e.g. finishBundle can produce output) so this would require adding `startBundleForGetInitialRestriction`, `finishBundleForGetInitialRestriction`, `startBundleForPairWithRestriction` and `finishBundleForPairWithRestriction`. I could see value in this for per bundle object lifetime management but any such change should likely happen outside of the scope of this PR.

Any reason not to use `setup`/`teardown` for your object cache?

Reply:
I just got the idea of `setup` /`teardown`. Thanks!